lib/benchpark/experiment.py

```diff
@@ -65,30 +114,124 @@ def compute_config_section(self):
+        self.excludes.append(exclude_clause)
 
     def compute_applications_section(self):
-        # TODO: is there some reasonable default?
         raise NotImplementedError(
             "Each experiment must implement compute_applications_section"
         )
 
+    def compute_applications_section_wrapper(self):
+        self.expr_name = []
+        self.variables = {}
+        self.zips = {}
+        self.matrix = []
+        self.excludes = []
+
+        self.compute_applications_section()
+
+        expr_name_list = [self.name, self.workload]
+        for cls in self.helpers:
+            helper_prefix = cls.get_helper_name_prefix()
+            if helper_prefix:
+                expr_name_list.append(helper_prefix)
+        expr_name = "_".join(expr_name_list + self.expr_name)
+
+        return {
+            self.name: {
+                "workloads": {
+                    self.workload: {
+                        "experiments": {
+                            expr_name: {
+                                "variants": {"package_manager": "spack"}, 
+                                "variables": self.variables,
+                                "zips": self.zips,
+                                "matrix": self.matrix,
+                                "exclude": {"where" : self.excludes} if self.excludes else {},
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+    def add_spack_spec(self, package_name, spec=None):
+        if spec:
+            self.package_specs[package_name] = {
+                "pkg_spec": spec[0],
+                "compiler": spec[1],
+            }
+        else:
+            self.package_specs[package_name] = {}
+
     def compute_spack_section(self):
-        # TODO: is there some reasonable default based on known variable names?
         raise NotImplementedError(
             "Each experiment must implement compute_spack_section"
         )
 
+    def compute_spack_section_wrapper(self):
+        for cls in self.helpers:
+            cls_package_specs = cls.compute_spack_section()
+            if (
+                cls_package_specs
+                and "packages" in cls_package_specs
+            ):
+                self.package_specs |= cls_package_specs["packages"]
+
+        self.compute_spack_section()
+
+        if not self.name in self.package_specs:
+             raise BenchparkError(
+                 f"Spack section must be defined for application package {self.name}"
+             )
+
+        spack_variants = [cls.get_spack_variants() for cls in self.helpers]
+        self.package_specs[self.name]["pkg_spec"] += " ".join(spack_variants+list(self.spec.variants["extra_spack_specs"])).strip()
+
+        return {
+            "packages": {k: v for k, v in self.package_specs.items() if v},
+            "environments": {self.name: {"packages": list(self.package_specs.keys())}},
+        }
+
```

Comment:
@scheibelp @pearce8 Refactored lib/experiment.py to remove much of the boilerplate code from the application experiment.py